File: src/utils/taxTest.ts
import type { Position } from '../types'
import { isClosedLot } from './rowDerivation'

/**
 * Czech 3-year time test (časový test): a gain on securities held more than three years is
 * exempt. Estimate only — see the TaxPanel footer. Mirrored by core/calc/TaxTest.kt against
 * test-fixtures/tax/cases.json.
 */

const DAY_MS = 86_400_000
const parse = (d: string) => Date.UTC(+d.slice(0, 4), +d.slice(5, 7) - 1, +d.slice(8, 10))
const iso = (ms: number) => new Date(ms).toISOString().slice(0, 10)

/** First date a sale is exempt: buy date + 3 years (29 Feb clamps to 28 Feb) + 1 day. */
export function timeTestDate(buyDate: string): string {
  const y = +buyDate.slice(0, 4) + 3
  const m = +buyDate.slice(5, 7) - 1
  const lastDay = new Date(Date.UTC(y, m + 1, 0)).getUTCDate()
  const d = Math.min(+buyDate.slice(8, 10), lastDay)
  return iso(Date.UTC(y, m, d) + DAY_MS)
}

export type TaxKind = 'exempt' | 'taxable' | 'pending'
export interface LotTaxStatus { kind: TaxKind; freeFrom: string; daysLeft: number }

export function lotTaxStatus(lot: Pick<Position, 'buyDate' | 'sellDate' | 'sellPrice'>, today: string): LotTaxStatus {
  const freeFrom = timeTestDate(lot.buyDate)
  if (lot.sellPrice != null && lot.sellDate) {
    return { kind: lot.sellDate >= freeFrom ? 'exempt' : 'taxable', freeFrom, daysLeft: 0 }
  }
  const daysLeft = Math.max(0, Math.round((parse(freeFrom) - parse(today)) / DAY_MS))
  return { kind: daysLeft === 0 ? 'exempt' : 'pending', freeFrom, daysLeft }
}

export interface YearSummary {
  year: number
  proceedsTaxable: number
  gainTaxable: number
  gainExempt: number
  /** Gross proceeds of taxable sales stay within the 100 000 CZK annual small-sales limit. */
  underSmallProceedsLimit: boolean
}

export const SMALL_PROCEEDS_LIMIT_CZK = 100_000

/** Realized gains per sell year, in CZK at each trade's own date. Newest year first. */
export function realizedByYear(
  positions: Position[],
  czkAt: (amount: number, currency: string, date: string) => number,
): YearSummary[] {
  const years = new Map<number, YearSummary>()
  positions.filter(isClosedLot).forEach((lot) => {
    const year = +lot.sellDate!.slice(0, 4)
    const y = years.get(year) ?? { year, proceedsTaxable: 0, gainTaxable: 0, gainExempt: 0, underSmallProceedsLimit: true }
    const proceeds = czkAt(lot.sellPrice! * lot.quantity, lot.currency, lot.sellDate!)
    const gain = proceeds - czkAt(lot.buyPrice * lot.quantity, lot.currency, lot.buyDate)
    if (lotTaxStatus(lot, lot.sellDate!).kind === 'exempt') y.gainExempt += gain
    else {
      y.gainTaxable += gain
      y.proceedsTaxable += proceeds
    }
    years.set(year, y)
  })
  return [...years.values()]
    .map((y) => ({ ...y, underSmallProceedsLimit: y.proceedsTaxable <= SMALL_PROCEEDS_LIMIT_CZK }))
    .sort((a, b) => b.year - a.year)
}

File: android/core/calc/src/main/kotlin/com/stocktracker/core/calc/TaxTest.kt
package com.stocktracker.core.calc

import com.stocktracker.core.model.Position
import java.time.LocalDate
import java.time.temporal.ChronoUnit

/**
 * Czech 3-year time test (časový test): a gain on securities held more than three years is
 * exempt. Estimate only. Mirrors src/utils/taxTest.ts against test-fixtures/tax/cases.json.
 */

/** First date a sale is exempt: buy date + 3 years (29 Feb clamps to 28 Feb) + 1 day. */
fun timeTestDate(buyDate: String): String = LocalDate.parse(buyDate).plusYears(3).plusDays(1).toString()

enum class TaxKind { EXEMPT, TAXABLE, PENDING }

data class LotTaxStatus(val kind: TaxKind, val freeFrom: String, val daysLeft: Long)

fun lotTaxStatus(buyDate: String, sellDate: String?, sellPrice: Double?, today: String): LotTaxStatus {
    val freeFrom = timeTestDate(buyDate)
    if (sellPrice != null && !sellDate.isNullOrEmpty()) {
        return LotTaxStatus(if (sellDate >= freeFrom) TaxKind.EXEMPT else TaxKind.TAXABLE, freeFrom, 0)
    }
    val daysLeft = maxOf(0L, ChronoUnit.DAYS.between(LocalDate.parse(today), LocalDate.parse(freeFrom)))
    return LotTaxStatus(if (daysLeft == 0L) TaxKind.EXEMPT else TaxKind.PENDING, freeFrom, daysLeft)
}

fun lotTaxStatus(lot: Position, today: String): LotTaxStatus = lotTaxStatus(lot.buyDate, lot.sellDate, lot.sellPrice, today)

data class YearSummary(
    val year: Int,
    val proceedsTaxable: Double,
    val gainTaxable: Double,
    val gainExempt: Double,
    /** Gross proceeds of taxable sales stay within the 100 000 CZK annual small-sales limit. */
    val underSmallProceedsLimit: Boolean,
)

const val SMALL_PROCEEDS_LIMIT_CZK = 100_000.0

/** Realized gains per sell year, in CZK at each trade's own date. Newest year first. */
fun realizedByYear(positions: List<Position>, czkAt: (Double, String, String) -> Double): List<YearSummary> =
    positions.filter(::isClosedLot)
        .groupBy { it.sellDate!!.take(4).toInt() }
        .map { (year, lots) ->
            var proceedsTaxable = 0.0
            var gainTaxable = 0.0
            var gainExempt = 0.0
            lots.forEach { lot ->
                val proceeds = czkAt(lot.sellPrice!! * lot.quantity, lot.currency, lot.sellDate!!)
                val gain = proceeds - czkAt(lot.buyPrice * lot.quantity, lot.currency, lot.buyDate)
                if (lotTaxStatus(lot, lot.sellDate!!).kind == TaxKind.EXEMPT) {
                    gainExempt += gain
                } else {
                    gainTaxable += gain
                    proceedsTaxable += proceeds
                }
            }
            YearSummary(year, proceedsTaxable, gainTaxable, gainExempt, proceedsTaxable <= SMALL_PROCEEDS_LIMIT_CZK)
        }
        .sortedByDescending { it.year }
